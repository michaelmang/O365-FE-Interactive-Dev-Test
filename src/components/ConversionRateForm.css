.container {
  display: flex;
  flex-direction: row;
  justify-content: center;
}

.field {
  margin-right: 20px;
}

.field:last-child {
  margin-right: 0;
}
