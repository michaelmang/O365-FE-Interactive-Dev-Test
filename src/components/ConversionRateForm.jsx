import React from 'react';

import styles from './ConversionRateForm.module.scss';
import Date from './Date.jsx';
import Select from './Select.jsx';

export default class ConversionRateForm extends React.PureComponent {
  state = {
    baseCurrency: '',
    date: '',
    targetCurrency: '',
  }

  handleBaseCurrency = ({ target }) => {
    this.setState({ baseCurrency: target.value });
  }

  handleDate = ({ target }) => {
    this.setState({ date: target.value });
  }

  handleTargetCurrency = ({ target }) => {
    this.setState({ targetCurrency: target.value });
  }

  render() {
    let {
      baseCurrency,
      date,
      targetCurrency,
    } = this.state;

    return (
      <form className={styles.container}>
        <Select
          label="Pick Base Currency"
          onSelect={this.handleBaseCurrency}
          value={baseCurrency}
        />
        <Select
          label="Pick Target Currency"
          onSelect={this.handleTargetCurrency}
          value={targetCurrency}
        />
        <Date onDate={this.handleDate} value={date} />
        <button className="button is-primary">Look Up</button>
      </form>
    );
  }
}
