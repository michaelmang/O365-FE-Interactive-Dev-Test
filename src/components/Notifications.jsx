import PropTypes from 'prop-types';
import React from 'react';

import styles from './Notifications.module.scss';

const Notifications = ({ errors }) => (
  <div className={styles.container}>
    {errors.map(renderError)}
  </div>
);

Notifications.propTypes = {
  errors: PropTypes.arrayOf(PropTypes.shape({
    message: PropTypes.string,
  })).isRequired,
};

const renderError = ({ message }, idx) => (
  <div key={idx} className="notification is-danger">
    {message}
  </div>
);

export default Notifications;
