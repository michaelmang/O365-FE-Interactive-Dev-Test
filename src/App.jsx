import 'bulma/css/bulma.css';
import React from 'react';

import styles from './App.module.scss';
import ConversionRateForm from './components/ConversionRateForm.jsx';
import Notifications from './components/Notifications.jsx';

export default class App extends React.PureComponent {
  state = {
    errors: [],
  }

  appendError = (error) => {
    this.setState({ errors: [error] }); // currently only support one error at a time
  }

  clearError = () => {
    this.setState({ errors: [] }); // effectively clears all errors until multiple errors supported
  };

  render() {
    let { errors } = this.state;
    return (
      <div className={styles.container}>
        <h1 className="title">Look Up Conversion Rate</h1>
        <ConversionRateForm
          appendError={this.appendError}
          clearError={this.clearError}
        />
        <Notifications errors={errors} />
      </div>
    );
  }
}
