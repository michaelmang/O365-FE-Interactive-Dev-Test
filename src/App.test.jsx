import { shallow } from 'enzyme';
import React from 'react';

import App from './App.jsx';
import ConversionRateForm from './components/ConversionRateForm.jsx';
import Notifications from './components/Notifications.jsx';
import renderSnapshot from './testing/renderSnapshot.js';

jest.mock('./App.module.scss', () => ({
  container: 'styles.container',
}));
jest.mock('./components/ConversionRateForm.jsx', () => 'ConversionRateForm');
jest.mock('./components/Notifications.jsx', () => 'Notifications');

describe('App', () => {
  it('matches the saved snapshot', () => {
    let { root: actual } = extractComponents(renderComponent());

    expect(renderSnapshot(actual)).toMatchSnapshot();
  });

  describe('errors appended', () => {
    it('passes down appended errors to Notifications', () => {
      let {
        conversionRateForm,
        notifications: subject,
        root,
      } = extractComponents(renderComponent());

      expect(subject).toHaveProp(
        'errors',
        [],
      );

      conversionRateForm.props().appendError({
        id: 'some-error',
        message: 'some error message',
      });
      root.update();

      subject = extractComponents(root).notifications;

      expect(subject).toHaveProp(
        'errors',
        [{ id: 'some-error', message: 'some error message' }],
      );
    });
  });

  describe('errors cleared', () => {
    it('passes down empty list of errors to Notifications when errors cleared', () => {
      let {
        conversionRateForm,
        notifications: subject,
        root,
      } = extractComponents(renderComponent());

      conversionRateForm.props().appendError({
        id: 'some-error',
        message: 'some error message',
      });
      root.update();

      subject = extractComponents(root).notifications;

      expect(subject).toHaveProp(
        'errors',
        [{ id: 'some-error', message: 'some error message' }],
      );

      conversionRateForm.props().clearError();
      root.update();


      subject = extractComponents(root).notifications;

      expect(subject).toHaveProp(
        'errors',
        [],
      );
    });
  });
  const extractComponents = (root) => ({
    conversionRateForm: root.find(ConversionRateForm),
    notifications: root.find(Notifications),
    root,
  });

  const renderComponent = (props) => (
    shallow(
      <App
        {...props}
      />
    )
  );
});
