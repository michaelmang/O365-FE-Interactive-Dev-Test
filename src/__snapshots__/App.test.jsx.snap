// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App matches the saved snapshot 1`] = `
<div
  className="styles.container"
>
  <h1
    className="title"
  >
    Look Up Conversion Rate
  </h1>
  <ConversionRateForm
    appendError={[Function]}
    clearError={[Function]}
  />
  <Notifications
    errors={Array []}
  />
</div>
`;
